Psalm Integration

timdev opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Psalm Integration, as requested in #3

timdev committed:
Introduce vimeo/psalm + psalm/plugin-phpunit as dev-depdencies.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Generate psalm-baseline.xml and add 'static-analysis' composer-script.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Psalm autofixes.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Remove redundnat doblock return-type declaration.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
change docblock type from union to intersection

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Add psalm typing to private FlashMessageMiddleware::$flashMessageFactory property:

This feels like this should be unnecessary, but there may be a bug in psalm about string-literal method names in array-style callables. https://github.com/vimeo/psalm/issues/6230

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Fix most psalm errors.

Signed-off-by: Tim Lieberman <dev@timdev.com>

froschdesign commented on src/FlashMessages.php:
```diff
@@ -155,15 +159,17 @@ public function prolongFlash(): void
     }
 
-    public function prepareMessages(SessionInterface $session, string $sessionKey): void
+    public function prepareMessages(SessionInterface $session): void
```
Caution: this class is not final and this is a change of the public API.

timdev replied:
Oh... drat.

timdev committed:
Suppress phpcs for test method with a very long name.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Restore method signature for FlashMessages::prepareMessages to fix BC.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev commented:
Closing for now, will reopen once I properly fix the BC issue I created.

boesing commented on psalm.xml:
```diff
@@ -0,0 +1,33 @@
+        </ignoreFiles>
+    </projectFiles>
+
+    <issueHandlers>
```
these issue handlers can be deleted - I sadly did not updated all the issues I've created last year. sorry for that.

boesing commented on psalm.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<psalm
+        totallyTyped="true"
```
should be replaced with `errorLevel="1"` as `totallyTyped` is deprecated since quite a while now

boesing commented:
You can switch the PR to draft mode. Done that for you. 👍🏼

boesing commented on psalm-baseline.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
If you manage to fix these two errors, we can avoid a baseline file in this project which would be awesome. 👍🏼

boesing replied:
Maybe casting a non-native typehinted "string" argument to string might help to tell psalm whats going on 👍🏼

timdev replied:
The one in the test can be (and has been) fixed by removing that test/merging it into the dependent test. Does that seem reasonable?

The one in `FlashMessages` is trickier. It's not about not recognizing a string. Rather, psalm really doesn't like `mixed`, but the public API here explicitly accepts `mixed $value`. 

With the caveat that I'm pretty new to psalm, it seems like psalm considers `mixed` to be a kind of abomination to be exorcised, and not just a maximally-wide type (like `any` in typescript).  In this instance, psalm knows full well (according to `@psalm-trace` if you you add some) that we're assigning `mixed` to `mixed`. But it appears psalm seems to take a hard line that you shouldn't do that. See: https://psalm.dev/r/69768c9c2a

It's equally (if not more) possible that I'm missing something, and I share your dream of not needing a baseline, so I'm going to give others with more psalm-experience a change to speak up.

timdev replied:
Per our brief discussion in slack, I've suppressed this error, as it's unavoidable.

timdev committed:
Clean/modernize psalm.xml

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Remove test method made redundant by type system, mostly because psalm complained abou tit.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Mark nullable argument as such.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Rename private method and related psalm-type.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Suppress unavoidable MixedAssignment psalm error.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Remove psalm-baseline.

Signed-off-by: Tim Lieberman <dev@timdev.com>

boesing commented on src/FlashMessages.php:
```diff
@@ -161,11 +165,17 @@ public function prepareMessages(SessionInterface $session, string $sessionKey):
         foreach ($sessionMessages as $key => $data) {
+            /**
+             * The public API of flash() and flashNow() explicitly allow calling
+             * code to pass `$value`s of any time, making this unavoidable.
```
guess you mean `type` rather than `time`? 😄

timdev replied:
Indeed.

timdev committed:
typofix

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Replace ::class reference with string-literal in ConfigProvider:

Psalm now gets upset about the class constant reference on a class that doesn't
exist. Swapping in a literal string here is fine, as the Expressive is not
expected to ever change.

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Ignore phpcs suggestion to use ::class

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev commented:
I've rebased this, and maneuvered around a new psalm error that cropped up in the meantime (see 2823f943).

Ocramius commented on src/FlashMessageMiddleware.php:
```diff
@@ -22,7 +22,10 @@ class FlashMessageMiddleware implements MiddlewareInterface
-    /** @var callable */
+    /**
+     * @var callable
+     * @psalm-var callable(SessionInterface, string):array | array{ 0: class-string, 1: string }
```
This union type seems borked - would be better to add a suppression on the assignment, than to mis-document this 🤔

timdev replied:
I struggled with this. For some reason, psalm [doesn't like](https://github.com/vimeo/psalm/issues/6230) string literals in the second element of an array-style callable. I've removed the union and sacrificed a chicken via an explicit @psalm-var in the method body.

timdev replied:
I've further improved this after @orklah helped me realize I was testing callability of the factory in a roundabout way.

Ocramius commented on src/FlashMessages.php:
```diff
@@ -161,11 +165,17 @@ public function prepareMessages(SessionInterface $session, string $sessionKey):
-        $sessionMessages = ! is_array($sessionMessages) ? [] : $sessionMessages;
+        $sessionMessages = $this->getStoredMessages($sessionKey);
 
+        /** @var array<string,mixed> $currentMessages */
```
Are messages generally `mixed`?

timdev replied:
While I just store strings there like a normal person, mezzio-flash doesn't impose any limitations. It's possible some madman is out there abusing that fact mightily :-)

Ocramius commented on src/FlashMessages.php:
```diff
@@ -183,4 +193,14 @@ public function prepareMessages(SessionInterface $session, string $sessionKey):
+    {
+        /** @var StoredMessages|null $messages */
+        $messages = $this->session->get($sessionKey ?? $this->sessionKey, []);
+        return ! is_array($messages) ? [] : $messages;
```
Unsure if relevant, but we may return `$message ?? []` here

timdev replied:
Nice catch.

Ocramius reviewed (approved):
IMO good improvement - up to you if you want to adjust the last rough edges, or if we should 🚢  as-is :)

timdev committed:
Tighten some screws per @ocramius

Signed-off-by: Tim Lieberman <dev@timdev.com>

timdev committed:
Test factory callability more simply, so psalm understands what's happening

Signed-off-by: Tim Lieberman <dev@timdev.com>

Ocramius commented:
This went completely under my radar. I rebased locally, and am now reviewing/merging. Sorry for the delay, @timdev!

Ocramius reviewed (approved):
:ship: thanks @timdev!
